[export] Serialize symbolic values

angelayi opened the pull request:
* Modified the SymInt schema to also store the hint of the SymInt if it is represented as a symbol so that when we reconstruct the SymInt, the hint will also exist on the node.
* GraphModuleDeserializer.deserialize now also optionally map of symbol names to range.

ReplaceSymSizeOpPass should not be needed after https://github.com/pytorch/pytorch/pull/103107 lands

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103273

angelayi committed:
[export] Serialize shape

[ghstack-poisoned]

angelayi committed:
Update on "[export] Serialize shape"




[ghstack-poisoned]

avikchaudhuri commented on torch/_export/serde/serialize.py:
```diff
@@ -624,10 +668,15 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
         return (
             ExportedProgram(
                 graph_module=serialized_graph_module,
-                opset_version=self.opset_version
```
how did this pass lint, lol...don't we insist on trailing commas?

angelayi replied:
I don't think the OSS linter insists on having trailing commas

avikchaudhuri commented:
Requesting changes mainly because I'm unsure of the motivation behind adding extra params to graph module deserializer.

angelayi commented on test/export/test_serialize.py:
```diff
@@ -159,7 +187,37 @@ def check_graph(self, fn, inputs) -> None:
 
         for orig, loaded in zip(flat_orig_outputs, flat_loaded_outputs):
-            self.assertTrue(torch.allclose(orig, loaded))
+            if isinstance(orig, torch.Tensor) and isinstance(loaded, torch.Tensor):
```
If they are tensors then I have to use `torch.allclose` to check equality.

avikchaudhuri replied:
Yeah I got that, just commenting on the double `isinstance` check.

angelayi commented on torch/_export/serde/serialize.py:
```diff
@@ -636,45 +685,82 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
-        self, serialized_graph_module: GraphModule,
+        self,
+        serialized_graph_module: GraphModule,
+        shape_env: Optional[symbolic_shapes.ShapeEnv] = None,
```
The `deserialize_range_constraints` function is called before deserializing the graph module, and I had it populate the shape_env with the symbols that appeared in those constraints first. 

This is under the assumption that most symbols that appear in the graph will also appear in the constraints (I want to say all symbols? but I'm not 100% sure on that although I don't have a counterexample). So I can first create all of the symbols in the shape_env, and then pass the `shape_env` and `symbol_name_to_symbol` list to the `GraphModuleDeserializer`. 

We need to populate `shape_env` with the ranges before tracing because if there's some data-dependent operation going on, like using a symint as a shape, deserializing the tensor metadata through the `torch.empty_strided` call will check on the ranges of the symint.

The `symbol_name_to_symbol` is also handy because when we're deserializing symint expressions, the `sympy.sympify` function uses it to map the same expressions to the same symbol (ex. if we have `symbol_name_to_symbol = {"s0", s0}`, whenever we see `"s0"` this will point to the same `s0`, and if we see something like `"7*s0"` this will also point to the same symbol `s0`.

avikchaudhuri replied:
Ah, populated by side effect: missed that.

I think `constraints` only contain input dimension symbols, right? Whereas `shape_env` can contain data-dependent ones, I suppose.

I suppose `shape_env` better contain all symbols appearing in the graph. These should be a superset of things in `constraints`, I think. And `symbol_name_to_symbol` should have all symbols in the graph as well...?

Is there a way to assert on this?

Also, I still don't fully understand whether you actually expect external users to call graph module deserializer directly. Why not make it private, remove the optional / default part, etc.?

zhxchen17 commented on torch/_export/serde/schema.py:
```diff
@@ -72,7 +72,7 @@ class Device:
 @dataclass
 class SymInt(_Union):
-    as_symbol: str
+    as_symbol: Tuple[str, int]
```
Why we don't go with the side table as discussed before? Are you going to change this in a later PR?

zhxchen17 commented on torch/_export/serde/serialize.py:
```diff
@@ -636,45 +685,82 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
-    def deserialize_sym(
-        self, s: Union[SymInt, SymBool]
-    ) -> Union[int, torch.SymInt, bool, torch.SymBool]:
+    def deserialize_sym_int(self, s: SymInt) -> Union[int, torch.SymInt]:
```
I guess you are handling unbacked symint in a later PR?

zhxchen17 reviewed (commented):
some initial comments

angelayi committed:
Update on "[export] Serialize symbolic values"

* Serialized range constraints and inline constraints into symbols represented as strings
* These symbols are then reconstructed in the deserialize_sym_int/bool functions by adding them to the shape env along with the ranges.
* Modified the SymInt schema to also store the hint of the SymInt if it is represented as a symbol so that when we reconstruct the SymInt, the hint will also exist on the node.
* GraphModuleDeserializer.deserialize now also optionally takes in a shape env and map of symbols to names.

ReplaceSymSizeOpPass should not be needed after https://github.com/pytorch/pytorch/pull/103107 lands




[ghstack-poisoned]

angelayi committed:
Update on "[export] Serialize symbolic values"

* Serialized range constraints and inline constraints into symbols represented as strings
* These symbols are then reconstructed in the deserialize_sym_int/bool functions by adding them to the shape env along with the ranges.
* Modified the SymInt schema to also store the hint of the SymInt if it is represented as a symbol so that when we reconstruct the SymInt, the hint will also exist on the node.
* GraphModuleDeserializer.deserialize now also optionally takes in a shape env and map of symbols to names.

ReplaceSymSizeOpPass should not be needed after https://github.com/pytorch/pytorch/pull/103107 lands




[ghstack-poisoned]

zhxchen17 commented on torch/_export/serde/serialize.py:
```diff
@@ -310,6 +305,59 @@ def deserialize_state_dict(serialized: bytes) -> Dict[str, torch.Tensor]:
+    )
+
+
+def _convert_int_to_sympy_int(val) -> sympy.Expr:
```
```suggestion
def _int_to_sympy_int(val) -> sympy.Expr:
```

zhxchen17 reviewed (commented):
I think we're good to go after 1 more iteration.

angelayi committed:
Update on "[export] Serialize symbolic values"

* Modified the SymInt schema to also store the hint of the SymInt if it is represented as a symbol so that when we reconstruct the SymInt, the hint will also exist on the node.
* GraphModuleDeserializer.deserialize now also optionally map of symbol names to range.

ReplaceSymSizeOpPass should not be needed after https://github.com/pytorch/pytorch/pull/103107 lands




[ghstack-poisoned]

angelayi commented on torch/_export/serde/serialize.py:
```diff
@@ -636,45 +689,86 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+                    self.symbol_name_to_symbol[val.expr_str] = sym
+
+                    if vr := self.symbol_name_to_range.get(val.expr_str):
+                        self.shape_env.var_to_range[sym] = symbolic_shapes.ValueRanges(
+                            vr.lower, vr.upper
+                        )
```
I can't because the constrain API doesn't take in a symbol as input 😅 but it uses a symbol to set the range...it's easier if I just set it directly...

angelayi committed:
Update on "[export] Serialize symbolic values"

* Modified the SymInt schema to also store the hint of the SymInt if it is represented as a symbol so that when we reconstruct the SymInt, the hint will also exist on the node.
* GraphModuleDeserializer.deserialize now also optionally map of symbol names to range.

ReplaceSymSizeOpPass should not be needed after https://github.com/pytorch/pytorch/pull/103107 lands




[ghstack-poisoned]

zhxchen17 commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -299,6 +299,14 @@ def guard_scalar(a):
     else:
         raise AssertionError(f"unrecognized scalar {a}")
 
+def _constrain_symbol_range(shape_env, s: sympy.Symbol, min: int, max: int):
```
nit, we already have "experimental" in the module path, no need for underscore here.
```suggestion
def constrain_symbol_range(shape_env, s: sympy.Symbol, min: int, max: int):
```

avikchaudhuri replied:
Eh, if we know this is going to be private, why not keep the underscore? One day symbolic shapes will no longer be experimental, and we don't have to do additional cleanup then...

avikchaudhuri replied:
Anyway, don't have to block on this comment, idc so much... :)

zhxchen17 replied:
I guess most of the functions in this file is meant to be private... yet they don't got underscore, so my point is mostly about consistency... but yeah don't feel blocked on this.

zhxchen17 reviewed (approved):
let's go

angelayi commented:
@pytorchbot merge
